### ISSUE TITLE
Issue when searching for strings with "API" in them

### ISSUE THREAD

**Issue by colombeen** [dnewsholme/PasswordState-Management#70]
When you're using Windows authentication and you search for *API*, the search will be translated to *winapi*. Example:

```powershell
PS C:\> Get-PasswordStatePassword -Search "sys_api_user" -Verbose
VERBOSE: using uri https://passwordstate/winapi/searchpasswords/?Search=sys_winapi_user
VERBOSE: GET https://passwordstate/winapi/searchpasswords/?Search=sys_winapi_user with 0-byte payload
The remote server returned an error: (404) Not Found.
At C:\Program Files\WindowsPowerShell\Mo...
```

You should replace "API" with "WINAPI" before you build the URI.

My suggestion would be to remove "/api" from the functions and add it in the *-PasswordStateResource functions

```powershell
Get-PasswordStateResource -URI "/passwords/?QueryAll"
# Get-PasswordStateResource does : ($passwordstateenvironment.baseuri) + "/API" or "/WINAPI" + $URI
# instead of replacing /api in $URI
```

**Comment by colombeen** [comment]
### Suggestion for "Get-PasswordStateResource"
```powershell
<#
.SYNOPSIS
   A function to simplify the Retrieval of password state resources via the rest API
.DESCRIPTION
       A function to simplify the Retrieval of password state resources via the rest API.
.EXAMPLE
    PS C:\> Get-PasswordStateResource -uri "/lists"
    Sets a password on the password api.
.PARAMETER URI
    The api resource to access such as /lists
.PARAMETER Method
    Optional Parameter to override the method from GET.
.OUTPUTS
    Will return the response from the rest API.
.PARAMETER ContentType
    Optional Parameter to override the default content type from application/json.
.PARAMETER ExtraParams
    Optional Parameter to allow extra parameters to be passed to invoke-restmethod. Should be passed as a hashtable.
.NOTES
    Daryl Newsholme 2018
#>
function Get-PasswordStateResource {
    [CmdletBinding()]
    param (
        [string]$uri,
        [string]$method = "GET",
        [string]$ContentType = "application/json",
        [hashtable]$extraparams = $null
    )

    begin {
        # Force TLS 1.2
        $SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol
        [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12
        # Import the environment
        $passwordstateenvironment = Get-PasswordStateEnvironment
		$api = "/api"
        # If the apikey is windowsauth then rebuild the uri string to match the windows auth apis, otherwise just build the api headers.
        Switch ($passwordstateenvironment.AuthType) {
            WindowsIntegrated {
                $api = "/winapi"
            }
            WindowsCustom {
                $api = "/winapi"
            }
            APIKey {
                switch -Wildcard ($uri){
                    /generatepassword* {
                        Write-Verbose "[$(Get-Date -format G)] Using generate password api key"
                        $headers = @{"APIKey" = "$($passwordstateenvironment.PasswordGeneratorAPIKey)"}
                    }
                    Default {
                        $headers = @{"APIKey" = "$($passwordstateenvironment.Apikey)"}
                    }
                }
            }
        }
    }

    process {
        $params = @{
            "UseBasicParsing" = $true
            "URI"             = "$($passwordstateenvironment.baseuri)$($api)$($uri)"
            "ContentType"     = $ContentType
            "Method"          = $method.ToUpper()
        }
        if (!$body){
            $params.Remove("Body")
        }
        if ($headers -and $null -ne $extraparams.Headers) {
            Write-Verbose "[$(Get-Date -format G)] Adding API Headers and extra param headers"
            $headers += $extraparams.headers
            $params += @{"headers" = $headers}
            $skipheaders = $true
        }
        if ($extraparams -and $null -eq $extraparams.Headers){
            Write-Verbose "[$(Get-Date -format G)] Adding extra parameter $($extraparams.keys) $($extraparams.values)"
            $params += $extraparams
        }

        if ($headers -and $skipheaders -ne $true) {
            Write-Verbose "[$(Get-Date -format G)] Adding API Headers only"
            $params += @{"headers" = $headers}
        }
        Switch ($passwordstateenvironment.AuthType) {
            APIKey {
                # Hit the API with the headers
                Write-Verbose "using uri $($params.uri)"
                $result = Invoke-RestMethod @params -TimeoutSec 60
            }
            WindowsCustom {
                Write-Verbose "using uri $($params.uri)"
                $result = Invoke-RestMethod @params -Credential $passwordstateenvironment.apikey -TimeoutSec 60
            }
            WindowsIntegrated {
                # Hit the api with windows auth
                Write-Verbose "using uri $($params.uri)"
                $result = Invoke-RestMethod @params -UseDefaultCredentials -TimeoutSec 60
            }
        }
    }

    end {
	    [System.Net.ServicePointManager]::SecurityProtocol = $SecurityProtocol
        return $result
    }
}
```

**Comment by dnewsholme** [comment]
I've got an easier fix which results in less code change simply changing the replace on "api" to replace on "/api/" instead. I'll push up the changes in a moment.

**Comment by colombeen** [comment]
It's true, your suggestion is less code, but a replace on a string that uses user input is always a risk

**Comment by dnewsholme** [comment]
You're right it is less safe but i'm regex matching against the beginning of a string only which should mean it only replaces the /api/ at the beginning.

**Comment by dnewsholme** [comment]
Fixed in #72  Build 4.3.0
